pytensor/tensor/elemwise.py

```diff
@@ -303,6 +303,80 @@ def process(self, r, pstate):
 pprint.assign(DimShuffle, DimShufflePrinter())
 
 
+def broadcast_shapes(
+    shapes: Sequence[Tuple[Optional[int], ...]],
+    broadcast_patterns: Sequence[Tuple[bool, ...]],
+):
+    """Broadcast shapes or raise an error.
+
+    This function already assumes all shapes have equal length, e.g. after Dimsuffle.
+
+    Parameters
+    ----------
+    shapes : Sequence[Tuple[Optional[int], ...]]
+        input shapes to broadcast
+    broadcast_patterns : Sequence[Tuple[bool,...]]
+        input broadcasting pattern constraints
+
+    Returns
+    -------
+    Tuple[Optional[int], ...], Tuple[bool,...]
+        shape and broadcasting pattern
+    """
+
+    def get_most_specialized_shape(shapes, bcasting):
+        seen_dims = set(zip(shapes, bcasting))
+        if len(seen_dims) == 1:
+            return next(iter(seen_dims))[0]
+        elif len(seen_dims) == 2 and (1, True) in seen_dims:
+            # this is fine since one dimension broadcasts to another common dimension
+            seen_dims.discard((1, True))
+            return next(iter(seen_dims))[0]
+        # we have set length >= 2 and it is not the case (1, True) in seen_dims
+        # let's drops dims that do not matter in comparison
+        # do not care about 1 that broadcasts
+        # the simple case was checked above so two discard
+        # never manage to produce an empty set
+        seen_dims.discard((1, True))
+        # do not care about unknown that does not broadcast because it
+        # should anyway match to known that did not broadcast
+        seen_dims.discard((None, False))
+        if len(seen_dims) > 1:
+            # we did not manage to specialize dims, raise an error
+            raise ValueError(f"shapes and broadcast mismatch: {seen_dims}")
+        return next(iter(seen_dims))[0]
+
+    # it is multiplied by nout because Elemwise supports multiple outputs
+    # (nout of them)
+    if len(shapes) < 1 or len(shapes) != len(broadcast_patterns):
+        raise ValueError(
+            "Length of arguments should be more than one and equal for shapes and broadcast patterns"
+        )
+    common_len = len(shapes[0])
+    if any(len(s) != common_len for s in shapes + broadcast_patterns):
+        raise ValueError("Shapes length mismatch, all lengths should be the same")
+    try:
+        out_shape = tuple(
+            get_most_specialized_shape(shape, bcasting)
+            for shape, bcasting in zip(
+                zip(*shapes),
+                zip(*broadcast_patterns),
+            )
+        )
+        out_broadcastable = tuple(all(bcast) for bcast in zip(*broadcast_patterns))
+    except ValueError as e:
+        raise ValueError(
+            "Incompatible Elemwise input broadcasting pattern: "
```

Comment:
How about something like "Incompatible Elemwise input broadcasting: Broadcasting is only allowed if the shape of the broadcasted axis is statically known to be one. Use `input.specify_shape` to inform pytensor that a shape is 1."
I don't think we have to explain why we do it like this in the error message. We could also add a faq entry, and link to that.